_stbt/core.py

```diff
@@ -1531,12 +1545,13 @@ def push_sample(self, sample):
 
     def on_error(self, pipeline, _bus, message):
         assert message.type == Gst.MessageType.ERROR
-        Gst.debug_bin_to_dot_file_with_ts(
-            pipeline, Gst.DebugGraphDetails.ALL, "ERROR")
+        if pipeline is not None:
+            # If we receive an error after the pipeline has been torn down:
+            Gst.debug_bin_to_dot_file_with_ts(
+                pipeline, Gst.DebugGraphDetails.ALL, "ERROR")
         err, dbg = message.parse_error()
         self.tell_user_thread(
             UITestError("%s: %s\n%s\n" % (err, err.message, dbg)))
-        _mainloop.quit()
```

Comment:
Not sure.  An error doesn't necessarily mean that data will stop flowing, but that's the convention.  I'm not quite sure what the `_mainloop.quit()` was achieving in here.  I guess it would avoid the possibility of a 10s wait at teardown.

I guess the sequence of events is:

1. Error occurs
2. User thread tries to pull a frame
3. Exception is raised in user thread
4. `self.teardown` is called

And then before this change:

5. Emit end of stream
6. Destroy pipeline

After this change:

5. Emit (and check) end of stream
6. Wait for EOS for up to 10s.
7. Set state NULL
8. Destroy pipeline.

I think the new implementation is probably an improvement.  We'd have to test with some real instances of ERRORs to be sure.